### ISSUE TITLE
Wrong type on extend parent class object field

### ISSUE THREAD

**Issue by metauro** [microsoft/TypeScript#33899]
**TypeScript Version:**  3.6.3

**Search Terms:**

**Code**

```ts
class Par {
  obj = {
    a: 'par property'
  }

  tempObj = {
    a: 'temp par property'
  }
}

class Sub extends Par {
  // error type, obj is any
  obj = Object.assign({
    b: 'sub property'
  }, this.obj as Par["obj"])

  _tempObj: any = this.tempObj

  // correct type, tempObj is Par["tempObj"] & {b : string}
  tempObj = Object.assign({
    b: 'temp sub property'
  }, this._tempObj as Par["tempObj"])  
}
```

**Expected behavior:**

```obj``` and ```tempObj``` type should be same

**Actual behavior:**

```obj``` type is any

**Playground Link:** 
http://www.typescriptlang.org/play/?ssl=17&ssc=31&pln=18&pc=1#code/MYGwhgzhAEAKYCdoG8CwAoa0D2AjAVtALwoZZZgBc0A5AA6LR0LZ0CmCALgJ41nQBfDP05sAtnQDyBYqUzkqtURKaNmrDjz7yh6XRlCQYAZQCuuaGwAeogHYATGPCRp5AejfQAlreDYECGzAnNA87Px4hCTS+EGcAHRGXgDmtgAUruTQuNQ0EOZMLOxcvPwCADShABZeEPGR0JBwiADaAESRbQC6AJTC8gD6ylIE1GC23LKcNXXDMf1YHtB+AXGh3OHyczLRBHGJUCnpmeQ5SuJ00PkW6sVaZZXTtfFDFzGNTq1t2-jdPVgYfToPy2CDYEBseIgbDJNK2NgAd2gZlwaR6-yAA

**Related Issues:**

**Comment by RyanCavanaugh** [comment]
Not mentioned in OP: this is flagged as an error.

Due to architectural constraints around how trees are marked as checked, this has to be treated as a circular reference despite the type assertion.

**Comment by metauro** [comment]
@RyanCavanaugh thanks for replying, it is possible extend parent class object field without temp variable?
